src/shogun/mathematics/linalg/LinalgBackendViennacl.h

```diff
@@ -0,0 +1,97 @@
+#include <shogun/mathematics/linalg/LinalgBackendBase.h>
+
+#ifndef Linalg_Backend_ViennaCL_H__
+#define Linalg_Backend_ViennaCL_H__
+
+#ifdef HAVE_CXX11
+#ifdef HAVE_VIENNACL
+
+#include <viennacl/vector.hpp>
+#include <viennacl/linalg/inner_prod.hpp>
+#include <shogun/mathematics/linalg/GPUMemoryViennaCL.h>
+
+namespace shogun
+{
+
+class LinalgBackendViennaCL : public LinalgBackendBase
+{
+	template <typename T>
+	friend struct GPUMemoryViennaCL;
+
+public:
+	#define BACKEND_GENERIC_DOT(Type) \
+	virtual Type dot(const SGVector<Type>& a, const SGVector<Type>& b) \
+	{  \
+		return dot_impl(a, b);  \
+	}
+
+	BACKEND_GENERIC_DOT(float32_t);
+	BACKEND_GENERIC_DOT(float64_t);
+	BACKEND_GENERIC_DOT(int32_t);
+
+	#define BACKEND_GENERIC_TO_GPU(Type) \
+	virtual GPUMemoryBase<Type>* to_gpu(const SGVector<Type>& vector) const \
+	{  \
+		return to_gpu_impl(vector);  \
+	}
+
+	BACKEND_GENERIC_TO_GPU(float32_t);
+	BACKEND_GENERIC_TO_GPU(float64_t);
+	BACKEND_GENERIC_TO_GPU(int32_t);
+
+	#define BACKEND_GENERIC_FROM_GPU(Type) \
+	virtual void from_gpu(const SGVector<Type>& vector, Type* data) const \
+	{  \
+		return from_gpu_impl(vector, data);  \
+	}
+
+	BACKEND_GENERIC_FROM_GPU(float32_t);
+	BACKEND_GENERIC_FROM_GPU(float64_t);
+	BACKEND_GENERIC_FROM_GPU(int32_t);
+
+	/** @return object name */
+	virtual const char* get_name() const { return "ViennaCL"; }
+
+//private:
+	template <typename T>
+	T dot_impl(const SGVector<T>& a, const SGVector<T>& b) const
+	{
+
+		// we know that the memory is of Vienna type as it was done with this backend
+		// if the developer is YOLO and changes the gpu backend after a gpu transfer and calls dot with
+		// he DESERVES a crash
+		GPUMemoryViennaCL<T>* a_gpu=static_cast<GPUMemoryViennaCL<T>*>(a.gpu_vector.get());
+		GPUMemoryViennaCL<T>* b_gpu=static_cast<GPUMemoryViennaCL<T>*>(b.gpu_vector.get());
+
+		return viennacl::linalg::inner_prod(a_gpu->data(a.size()), b_gpu->data(b.size()));
+	}
+
+	template <typename T>
+	GPUMemoryBase<T>* to_gpu_impl(const SGVector<T>& vector) const \
+	{
+		GPUMemoryViennaCL<T>* gpu_vec;
+		gpu_vec = new GPUMemoryViennaCL<T>();
```

Comment:
Hey @vigsterkr  may I ask how/where/should this `gpu_vec` be deleted?
I **think** it will be freed by SGVector as the pointer is passed to SGVector constructor but I'm not sure..